src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java

```diff
@@ -202,6 +211,45 @@ public void shouldNotThrowExceptionsWhenAssertionsAreValid() {
         verify(hubSignatureValidator, times(2)).validate(any(), any());
     }
 
+    @Test
+    public void shouldExpectAListContainingAnAuthnAssertion() {
+        Assertion authnAssertion = anAuthnStatementAssertion(IdaAuthnContext.LEVEL_2_AUTHN_CTX, "requestId").buildUnencrypted();
+        List<Assertion> assertions = Arrays.asList(authnAssertion);
+
+        exception.expect(NullPointerException.class);
```

Comment:
Why does it expect a `NullPointerException.class`? and also why does the `nonMatchingAssertionService.translate(assertions);` happen afterwards?

Reply:
What this is supposed to mean is that a NullPointerException is expected to be thrown later in the test (i.e. when `nonMatchingAssertionService.translate(assertions)` is called).  That's meant to happen because an assertion is missing.  It *is* happening, but it looks to me like these tests are actually back to front: the one which is meant to test a missing MDSassertion instead tests a missing AuthnAssertion, and vice versa.

Part of the reason this is all so confusing is that I've been lazy and just let the code throw an ordinary NullPointerException; it would be clearer if a custom exception were thrown instead.

Reply:
So essentially these test are passing as is because the translate(assertions) is expecting 2 assertions instead of one. Otherwise the assertions shouldn't be interchangeable.

Reply:
Well, that's not quite true.  If you give it multiple MDSAssertions, it'll still NPE because it can't find an AuthnAssertion.  This probably needs to be made clearer in the test cases though.

Reply:
We should probably wrap this in a nicer exception so that it logs nicer or we use some static typing and make `validate` do all these checks. If `validate` returned a `ValidatedResponse` and `translate` took that validated response, we could rely on the assertions being correct (i.e. 1 MDS, 1 Authn) and use the type system to make sure nobody is calling `translate` on unvalidated assertions. This will be tracked as tech debt so no need to deal with this as part of this card.

Reply:
We'll deal with this when we have a card for it. Investigated wrapping this in a nicer exception though but turned out a little messy. So abandoned it.